### PR TITLE
(Closes #2261) Splitting and renaming DynLoop

### LINKED ISSUE #2261: Splitting and Renaming DynLoop from src/psyclone/dynamo0p3.py
sub-issue of https://github.com/stfc/PSyclone/issues/2235

### PR DESCRIPTION
**Pull request by mo-lottieturner**
closes #2261

### CODE REVIEW

**Comment by mo-lottieturner** [comment]
there are a couple of instances of `psyclone.psyGen.DynLoop` that I'm not sure should have been DynLoops in the first place? but have changed to LFRicLoop for consistency's sake for the moment

**Comment by mo-lottieturner** [comment]
@rupertford there's a line missing test coverage that mentions #1618 in a TODO - I see you closed that issue in July, are you happy for this section to be deleted (l779 in domain/lfric/lfric_loop.py on this branch, l7120 in dynamo0p3.py on master)

```
             if not asym:
                # TODO #1618: once the symbols are all defined,
                # this should not happen anymore.
                raise InternalError(f"No symbol for last_cell_all_colours"
                                    f"defined for kernel "
                                    f"'{self.kernel.name}'.")```

**Comment by rupertford** [comment]
Hi @mo-lottieturner, the reference to #1618 is certainly out-of-date. Looking at the implementation of `last_cell_all_colours_symbol` in `lfric_kern.py`, I see it is using `self.scope.symbol_table....` so it has been updated to use the PSyIR symbol table and therefore the offending code can be removed.

File: src/psyclone/domain/lfric/lfric_builtins.py [hunk 1]
```diff
@@ -128,7 +128,7 @@ def create(call, parent=None):
         # Create the loop over the appropriate entity.
         # Avoid circular import
         # pylint: disable=import-outside-toplevel
-        from psyclone.dynamo0p3 import DynLoop
+        from psyclone.domain.lfric import LFRicLoop
```
Comment (arporter) [inline src/psyclone/domain/lfric/lfric_builtins.py:131]: Can this be moved to the top if it is made more specific: `from psyclone.domain.lfric.lfric_loop import...`?

File: src/psyclone/domain/lfric/lfric_kern.py [hunk 2]
```diff
@@ -674,8 +674,8 @@ def gen_code(self, parent):
                            entity_decls=["cell"]))
         # Import here to avoid circular dependency
         # pylint: disable=import-outside-toplevel
-        from psyclone.dynamo0p3 import DynLoop
-        parent_loop = self.ancestor(DynLoop)
+        from psyclone.domain.lfric import LFRicLoop
```
Comment (arporter) [inline src/psyclone/domain/lfric/lfric_kern.py:677]: Again, can this be moved up?

File: src/psyclone/domain/lfric/lfric_kern_call_factory.py [hunk 3]
```diff
@@ -75,8 +75,8 @@ def create(call, parent=None):
             loop_type = ""
         # Import here to avoid circular dependency
         # pylint: disable=import-outside-toplevel
-        from psyclone.dynamo0p3 import DynLoop
-        cloop = DynLoop(parent=parent, loop_type=loop_type)
+        from psyclone.domain.lfric import LFRicLoop
```
Comment (arporter) [inline src/psyclone/domain/lfric/lfric_kern_call_factory.py:78]: Can this be moved up?

File: src/psyclone/domain/lfric/lfric_loop.py [hunk 4]
```diff
@@ -0,0 +1,1049 @@
...
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors R. W. Ford, A. R. Porter and S. Siso, STFC Daresbury Lab
+# Modified I. Kavcic, A. Coughtrie, L. Turner and O. Brunt, Met Office
+# Modified J. Henrichs, Bureau of Meteorology
+# Modified A. B. G. Chalk and N. Nobre, STFC Daresbury Lab
+
+''' This module implements the PSyclone LFRic API by specialising the Loop
```
Comment (arporter) [inline src/psyclone/domain/lfric/lfric_loop.py:39]: Not yours but s/Loop/PSyLoop/

File: src/psyclone/domain/lfric/lfric_loop.py [hunk 5]
```diff
@@ -0,0 +1,1049 @@
...
+    base class in psyGen.py
```
Comment (arporter) [inline src/psyclone/domain/lfric/lfric_loop.py:40]: s/in/from/ and full stop please.

File: src/psyclone/domain/lfric/lfric_loop.py [hunk 6]
```diff
@@ -0,0 +1,1049 @@
...
+     '''
```
Comment (arporter) [inline src/psyclone/domain/lfric/lfric_loop.py:41]: No leading whitespace here please.

**Review by arporter — CHANGES_REQUESTED** [review]
Nice job. A few little nitpicky things and some requests to see if you can hoist some imports but apart from that it's all good.
I'll fire off the integration tests now.

File: src/psyclone/domain/lfric/lfric_loop.py (hunk 4 above)
Comment (mo-lottieturner) [reply src/psyclone/domain/lfric/lfric_loop.py:39]: I'm not clear on what you're wanting me to correct here, could you rephrase/explain what the s/../.. thing means?

Comment (arporter) [reply src/psyclone/domain/lfric/lfric_loop.py:39]: Sorry, that's sed notation - it means replace "Loop" with "PSyLoop".

File: src/psyclone/domain/lfric/lfric_kern.py (hunk 2 above)
Comment (mo-lottieturner) [reply src/psyclone/domain/lfric/lfric_kern.py:677]: no - LFRicLoop uses LFRicKern and LFRicKern uses LFRicLoop

File: src/psyclone/domain/lfric/lfric_kern_call_factory.py (hunk 3 above)
Comment (mo-lottieturner) [reply src/psyclone/domain/lfric/lfric_kern_call_factory.py:78]: Moved!

File: src/psyclone/domain/lfric/lfric_loop.py (hunk 4 above)
Comment (mo-lottieturner) [reply src/psyclone/domain/lfric/lfric_loop.py:39]: done

**Comment by mo-lottieturner** [comment]
@arporter back to you

File: src/psyclone/domain/lfric/lfric_builtins.py (hunk 1 above)
Comment (mo-lottieturner) [reply src/psyclone/domain/lfric/lfric_builtins.py:131]: oops, missed this, will tackle after lunch

Comment (mo-lottieturner) [reply src/psyclone/domain/lfric/lfric_builtins.py:131]: no, LFRicLoop uses LFRicBuiltins and vice versa

**Review by arporter — APPROVED** [review]
All requested changes have been made (thanks).
Ref. guide builds fine.
I've brought the branch up to master and the CI and integration tests are all green.
Will proceed to merge.
